cloudtrust-common/src/main/java/io/cloudtrust/crypto/CryptoUtil.java

```diff
@@ -169,34 +205,82 @@ public static String computeHmacForDatabaseStorage(String input) {
+                if (keys.size() == 0) {
+                    throw new IllegalStateException("Cannot find any appropriate key from environment variable " +
+                            DB_ENCRYPTION_KEY_ENV_VAR_NAME);
                 }
-                DB_ENCRYPTION_KEY = Base64.getDecoder().decode(keyB64);
+                CURRENT_DB_ENCRYPTION_KEY = keys.get(0); // take the most recent key (top-most one)
+                HISTORY_DB_ENCRYPTION_KEY = keys.stream().collect(Collectors.toMap(KeyEntry::getKid, KeyEntry::getValue));
             }
-            return DB_ENCRYPTION_KEY;
+            return CURRENT_DB_ENCRYPTION_KEY;
         } catch (Exception ex) {
             throw new IllegalStateException("Cannot load the " + DB_ENCRYPTION_KEY_ENV_VAR_NAME, ex);
         }
     }
 
     private static byte[] getCtDatabaseHmacKey() {
         try {
-            if (DB_HMAC_KEY == null) {
-                String keyB64 = System.getenv(DB_HMAC_KEY_ENV_VAR_NAME); // the key is encoded in Base64
-                if (keyB64 == null) {
-                    throw new IllegalStateException("Cannot load the " + DB_HMAC_KEY_ENV_VAR_NAME);
+            if (CURRENT_DB_HMAC_KEY == null) {
+                List<KeyEntry> keys = loadKeysFromEnvironment(DB_HMAC_KEY_ENV_VAR_NAME);
+                if (keys.size() == 0) {
+                    throw new IllegalStateException("Cannot find an appropriate key from environment variable " +
+                            DB_HMAC_KEY_ENV_VAR_NAME);
                 }
-                DB_HMAC_KEY = Base64.getDecoder().decode(keyB64);
+                CURRENT_DB_HMAC_KEY = keys.get(0); // take the most recent key (top-most one)
+                HISTORY_DB_HMAC_KEY = keys.stream().collect(Collectors.toMap(KeyEntry::getKid, KeyEntry::getValue));
             }
-            return DB_HMAC_KEY;
+            return CURRENT_DB_HMAC_KEY.value;
         } catch (Exception ex) {
             throw new IllegalStateException("Cannot load the " + DB_HMAC_KEY_ENV_VAR_NAME, ex);
         }
     }
 
+    private static List<KeyEntry> loadKeysFromEnvironment(String envVariableName) throws JsonProcessingException {
+        String keys = System.getenv(envVariableName); // the key is encoded in Base64
+        if (keys == null) {
+            throw new IllegalStateException("Cannot load the environment variable" + envVariableName);
+        }
+        return new ObjectMapper().readValue(keys, new TypeReference<List<KeyEntry>>() {
+        });
+    }
+
+    static class KeyEntry {
+        @JsonProperty("kid")
+        String kid;
+        @JsonProperty("value")
+        byte[] value;
+
+        public String getKid() {
+            return kid;
+        }
+
+        public byte[] getValue() {
+            return value;
+        }
+    }
+
+    static class EncryptedData {
+        @JsonProperty("kid")
+        String kid;
+        @JsonProperty("val")
+        String val;
```

Comment:
I don't get you point here, since **val** in this case is anyway a base64-encoded string. I think you are thinking about the credential secretData enclosing an _EncryptedData_ object. But this should be handled in the credential-specific classes.
Am I missing something?